src/audio/base_fw.c

```diff
@@ -228,13 +235,99 @@ static int basefw_mem_state_info(uint32_t *data_offset, char *data)
 	return 0;
 }
 
+static uint32_t basefw_set_system_time(uint32_t param_id,
+				       bool first_block,
+				       bool last_block,
+				       uint32_t data_offset,
+				       char *data)
+{
+	if (!(first_block && last_block))
+		return IPC4_INVALID_REQUEST;
+
+	global_system_time_info.host_time.val_l = (*(struct ipc4_system_time *)(data)).val_l;
+	global_system_time_info.host_time.val_u = (*(struct ipc4_system_time *)(data)).val_u;
+	uint64_t current_dsp_time = sof_cycle_get_64();
+
+	global_system_time_info.dsp_time.val_l = (uint32_t)(current_dsp_time);
+	global_system_time_info.dsp_time.val_u = (uint32_t)(current_dsp_time >> 32);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_system_time(uint32_t *data_offset, char *data)
+{
+	struct ipc4_system_time *system_time = (struct ipc4_system_time *)data;
+
+	system_time->val_l = global_system_time_info.host_time.val_l;
+	system_time->val_u = global_system_time_info.host_time.val_u;
+	*data_offset = sizeof(struct ipc4_system_time);
+	return IPC4_SUCCESS;
+}
+
+static uint32_t basefw_get_ext_system_time(uint32_t *data_offset, char *data)
+{
+#if CONFIG_ACE_V1X_ART_COUNTER && CONFIG_ACE_V1X_RTC_COUNTER
+	struct ipc4_ext_system_time *ext_system_time = (struct ipc4_ext_system_time *)(data);
+	struct ipc4_ext_system_time ext_system_time_data = {0};
+
+	uint64_t host_time = ((uint64_t)global_system_time_info.host_time.val_u << 32)
+				| (uint64_t)global_system_time_info.host_time.val_l;
+	uint64_t dsp_time = ((uint64_t)global_system_time_info.dsp_time.val_u << 32)
+				| (uint64_t)global_system_time_info.dsp_time.val_l;
+	(void)host_time;
+	(void)dsp_time;
+
+	uint64_t art = 0;
+	uint64_t wallclk = 0;
+	uint64_t rtc = 0;
+
+	const struct device *dev = DEVICE_DT_GET(DT_NODELABEL(ace_art_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &art);
+
+	wallclk = sof_cycle_get_64();
+	ext_system_time_data.art_l = (uint32_t)art;
+	ext_system_time_data.art_u = (uint32_t)(art >> 32);
+	uint64_t delta = (wallclk - dsp_time) / (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC / 1000000);
+
+	uint64_t new_host_time = (host_time + delta);
+
+	ext_system_time_data.utc_l = (uint32_t)new_host_time;
+	ext_system_time_data.utc_u = (uint32_t)(new_host_time >> 32);
+
+	dev = DEVICE_DT_GET(DT_NODELABEL(ace_rtc_counter));
+
+	if (!dev) {
+		LOG_DBG("board: ACE V1X device binding failed");
+		return IPC4_FAILURE;
+	}
+
+	counter_get_value_64(dev, &rtc);
```

Comment:
Make sure this value is provided by RTC HAL function on Intel platform.

Reply:
@mmaka1 Please correct me if I'm wrong, but I think that the whole point of what we're doing here is to abandon those HAL functions in favor of Zephyr legitimate drivers. In case of RTC and ART, two additional Intel drivers have been added to Zephyr (PR: https://github.com/zephyrproject-rtos/zephyr/pull/49175). They are supposed to do the same thing as the HAL functions that you've mentioned. To ensure that we are getting the right values, we pass the the proper device as an argument of counter_get_value_64() function.

Reply:
Oh, I left my both comments regarding counter_get_value_64() API incomplete, sorry for that. I meant '... while working on other platforms too'. Are we OK with the current implementation where this IPC4 call fails on platforms other than Intel ACE?

Reply:
@mmaka1  That's a good point. So basicly the code to look up which device to use for counter_get_value_64(), should be platform specific, and ideally behind a bit of abstraction to make this scale when more platforms are added.

Considering this is the first platform to support this, and we return IPC4_UNAVAILABLE, I think this is still ok starting point and the generalization of the interface can be done in follow-up PRs. @mwasko @lgirdwood what do you think?

Reply:
This is using common api (counter_get_value_64) so already it is not platform specific, if the device didn't have ace in the name everything would be ok, just the implementation is platform specific but this is internal zephyr thing, like for dma we have platform specific implementations but one generic device with common api, not sure why there is ace everywhere but I dont know history of this driver.

Reply:
The history behind is that it was named ace following the naming convention of other Zephyr drivers (interrupt drivers to be more precise, which introduced the distinction between platforms). I agree that this was not the best choice, but we have discussed this within our team and the conclusion is that it does not matter that much at this point and for this PR. To avoid further confusion, those Zephyr  drivers will be renamed to intel_art_counter and intel_rtc_counter in the future.

Reply:
@abonislawski @piotrkmx It's not just the "ace" part. Code in "src/audio/" should be vendor neutral, so this should rely on just a generic counter device that is used in the IPC4 device counter (could be non-ART on non-Intel platforms). But alas, this is the first hardware that adds the support, so I agree this can be generalized in later PRs when/if the need comes. FYI @lgirdwood